### ISSUE TITLE
terraform-0.14 import of modules that use for_each

### ISSUE THREAD

**Issue by kimor79** [hashicorp/terraform#26764]
### Terraform Version

```
Terraform v0.14.0-beta2
+ provider registry.terraform.io/hashicorp/aws v3.13.0
+ provider registry.terraform.io/hashicorp/time v0.6.0
+ provider registry.terraform.io/splunk-terraform/signalfx v6.0.0
```

### Terraform Configuration Files

This is a very complex configuration. I've tried to narrow it down as much as possible, but the below is effectively untested

```terraform
locals {
  vpcs = {
    "12345-us-east-1" = {
      "account_id" = "12345",
      "cidr_block" = "192.168.10.0/24",
      "vpc_id" = "vpc-abcdef",
      "vpc_name" = "vpc1",
    },
    "45678-us-east-1" = {
      "account_id" = "45678",
      "cidr_block" = "192.168.11.0/24",
      "vpc_id" = "vpc-asdfg",
      "vpc_name" = "vpc2",
    },
  }

  peers = {
    "12345-us-east-1" = [
      "45678-us-east-1",
    ]
  }

  vpc_peering_initiates = merge(
    {
      for target in local.peers[local.account_region] : target => {
        "account_id" = local.vpcs[target]["account_id"],
        "cidr_block" = local.vpcs[target]["cidr_block"],
        "vpc_id"     = local.vpcs[target]["vpc_id"],
        "vpc_name"   = local.vpcs[target]["vpc_name"],
      }
    },
  )
}

module "vpc_peering_requester" {
  for_each = { for target, vpc_info in local.vpc_peering_initiates : target => vpc_info if target != local.account_region }

  source = "./vpc_peering_requester"

  peer_account_id = each.value["account_id"]
  peer_cidr_block = each.value["cidr_block"]
  peer_vpc_id     = each.value["vpc_id"]
  peer_vpc_name   = each.value["vpc_name"]
}
```

### Expected Behavior

`terraform import` should import successfully

### Actual Behavior

```
Error: Invalid for_each argument

  on path/to/my.tf line 2, in module "vpc_peering_requester":
   2:   for_each = { for target, vpc_info in local.vpc_peering_initiates : target => vpc_info if target != local.account_region }

The "for_each" value depends on resource attributes that cannot be determined
until apply, so Terraform cannot predict how many instances will be created.
To work around this, use the -target argument to first apply only the
resources that the for_each depends on.
```

### Steps to Reproduce

`terraform import "an unrelated module to the configuration above"`

### Additional Context

It's kind of hard to follow the chain of PRs and comments but I think this type of importing is supposedly fixed in tf-0.14 via https://github.com/hashicorp/terraform/pull/26490 but it doesn't work for me.

If I comment out the offending module (and anything depending on that), I can import successfully. But that doesn't help when the thing I need to import is for that module itself!

### References

- https://github.com/hashicorp/terraform/issues/26258
- https://github.com/hashicorp/terraform/pull/26490 possibly

**Comment by kimor79** [comment]
If I import something already imported, it does recognize that the resource has already been imported but still gives the _invalid for each_ error:
```
Error: Resource already managed by Terraform

Terraform is already managing a remote object for
"an unrelated module to the configuration above". To import to
this address you must first remove the existing object from the state.


Error: Invalid for_each argument

  on path/to/my.tf line 2, in module "vpc_peering_requester":
   2:   for_each = { for target, vpc_info in local.vpc_peering_initiates : target => vpc_info if target != local.account_region }

The "for_each" value depends on resource attributes that cannot be determined
until apply, so Terraform cannot predict how many instances will be created.
To work around this, use the -target argument to first apply only the
resources that the for_each depends on.
```

**Comment by jbardin** [comment]
Hi @kimor79 

I believe this should work with the 0.14 releases, but I'm not able to reproduce the problem without a complete configuration. The `local.account_region` value is not present here, and does effect the ability to evaluate the module block shown. Can you show how that value is derived in your configuration? 

The import process is somewhat limited in what it can do, and does require a working config, even for unrelated resources. You should be able to successfully run a plan before importing the resource, and have it only show changed related to the resource that needs to be created.

**Comment by kimor79** [comment]
```
data "aws_caller_identity" "current" {}
data "aws_region" "current" {}

locals {
  account_id = data.aws_caller_identity.current.account_id
  region     = data.aws_region.current.name

  account_region = "${local.account_id}-${local.region}"
}
```

plan runs successfully. From an empty state: `Plan: 229 to add, 0 to change, 0 to destroy.`

And if I comment out the offending code, import unrelated resources, uncomment the offending code: `Plan: 223 to add, 6 to change, 0 to destroy.`

On another set of similar/the same configs but without existing resources (different AWS accounts, etc) I can `apply` successfully

**Comment by jbardin** [comment]
Sorry if I wasn't clear with my suggestion -- have you done a complete `terraform apply` _without_ the resources that are going to be imported first? From the behavior, it looks like `local.account_id` and `local.region` are unknown when you're trying to import.

**Comment by kimor79** [comment]
Using the same config but a different role in the provider (aka in a different AWS account), yes I can apply successfully. In this account I cannot apply because the resources already exist, hence me trying to import them :)

Same error if I hardcode the account id and region instead of the _data_ call

**Comment by kimor79** [comment]
I've narrowed down the config:
```
# main.tf

locals {
  vpc_inputs = {
    "12345-us-east-1" = {
      "cidr_prefix"     = "10.1",
      "vpc_name_prefix" = "vpc-A",
    }
  }

  vpcs = {
    "12345-us-east-1" = module.account-12345-us-east-1.vpc,
    "56789-us-east-1" = {
      "account_id" = "56789",
      "cidr_block" = "192.168.0.0/24",
      "region"     = "us-east-1",
      "vpc_id"     = "vpc-ghijkl",
      "vpc_name"   = "vpc-B",
    }
  }

  vpc_peering_initiates = {
    "12345-us-east-1" = [
      "56789-us-east-1",
    ]
  }
}

module "account-12345-us-east-1" {
  source = "./account"

  providers = {
    aws = aws.account-12345-us-east-1
  }

  vpc_inputs            = local.vpc_inputs
  vpc_peering_initiates = local.vpc_peering_initiates
  vpcs                  = local.vpcs
}

provider "aws" {
  alias = "account-12345-us-east-1"
  assume_role {
    role_arn = "arn:aws:iam::12345:role/my-role"
  }
  region = "us-east-1"
}
```

```
# ./account/main.tf

provider "aws" {}

variable "vpc_inputs" {
  default = null
}

variable "vpc_peering_initiates" {
  default = null
}

variable "vpcs" {
}

data "aws_caller_identity" "current" {}
data "aws_region" "current" {}

locals {
  account_id = data.aws_caller_identity.current.account_id
  region     = data.aws_region.current.name

  account_region = "${local.account_id}-${local.region}"

  vpc_peering_initiates = {
    for target in var.vpc_peering_initiates[local.account_region] : target => {
      "account_id" = var.vpcs[target]["account_id"],
      "cidr_block" = var.vpcs[target]["cidr_block"],
      "vpc_id"     = var.vpcs[target]["vpc_id"],
      "vpc_name"   = var.vpcs[target]["vpc_name"],
    }
  }
}

module "vpc" {
  source = "../vpc"

  cidr_prefix     = var.vpc_inputs[local.account_region]["cidr_prefix"]
  vpc_name_prefix = var.vpc_inputs[local.account_region]["vpc_name_prefix"]
}

module "vpc_peering_requester" {
  for_each = { for target, vpc_info in local.vpc_peering_initiates : target => vpc_info if target != local.account_region }

  source = "../peering_requester"
}

output "vpc" {
  value = merge(module.vpc.info, {
    "account_id"     = local.account_id,
    "account_region" = local.account_region,
    "region"         = local.region,
  })
}
```

```
# ./vpc/main.tf

variable "cidr_prefix" {
  type = string
}

variable "vpc_name_prefix" {
  type = string
}

data "aws_region" "current" {}

locals {
  vpc_name = "${var.vpc_name_prefix}-${data.aws_region.current.name}"
}

resource "aws_vpc" "main" {
  cidr_block = "${var.cidr_prefix}.0.0/16"
}

output "info" {
  value = {
    "cidr_block" = aws_vpc.main.cidr_block,
    "vpc_id"     = aws_vpc.main.id,
    "vpc_name"   = local.vpc_name,
  }
}
```

and `./peering_requester/main.tf` is just an empty file

terraform plan:
```
$ terraform-0.14 plan

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # module.account-12345-us-east-1.module.vpc.aws_vpc.main will be created
  + resource "aws_vpc" "main" {
      + arn                              = (known after apply)
      + assign_generated_ipv6_cidr_block = false
      + cidr_block                       = "10.1.0.0/16"
      + default_network_acl_id           = (known after apply)
      + default_route_table_id           = (known after apply)
      + default_security_group_id        = (known after apply)
      + dhcp_options_id                  = (known after apply)
      + enable_classiclink               = (known after apply)
      + enable_classiclink_dns_support   = (known after apply)
      + enable_dns_hostnames             = (known after apply)
      + enable_dns_support               = true
      + id                               = (known after apply)
      + instance_tenancy                 = "default"
      + ipv6_association_id              = (known after apply)
      + ipv6_cidr_block                  = (known after apply)
      + main_route_table_id              = (known after apply)
      + owner_id                         = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Changes to Outputs:
  + info = {
      + cidr_block = "10.1.0.0/16"
      + vpc_id     = (known after apply)
      + vpc_name   = "vpc-A-us-east-1"
    }
  + vpc  = {
      + account_id     = "12345"
      + account_region = "12345-us-east-1"
      + cidr_block     = "10.1.0.0/16"
      + region         = "us-east-1"
      + vpc_id         = (known after apply)
      + vpc_name       = "vpc-A-us-east-1"
    }

------------------------------------------------------------------------

Note: You didn't specify an "-out" parameter to save this plan, so Terraform
can't guarantee that exactly these actions will be performed if
"terraform apply" is subsequently run.
```

terraform import
```
$ terraform-0.14 import module.account-12345-us-east-1.module.vpc.aws_vpc.main vpc-abcdef
module.account-12345-us-east-1.module.vpc.aws_vpc.main: Importing from ID "vpc-abcdef"...
module.account-12345-us-east-1.module.vpc.aws_vpc.main: Import prepared!
  Prepared aws_vpc for import
module.account-12345-us-east-1.module.vpc.aws_vpc.main: Refreshing state... [id=vpc-abcdef]

Error: Invalid for_each argument

  on account/main.tf line 41, in module "vpc_peering_requester":
  41:   for_each = { for target, vpc_info in local.vpc_peering_initiates : target => vpc_info if target != local.account_region }

The "for_each" value depends on resource attributes that cannot be determined
until apply, so Terraform cannot predict how many instances will be created.
To work around this, use the -target argument to first apply only the
resources that the for_each depends on.
```

**Comment by jbardin** [comment]
Hi @kimor79,

Thanks for the production. I verified this is now working in the latest release.
